TEST: Added Unit test for test student cannot put teaching period

Guoheng-Liao opened the pull request:
# Description

Modified the variables in setting_test.rb and unified the variable format.

Added unit test for activity type error method.

Added unit test for test student cannot put teaching period.



# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation if appropriate
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have created or extended unit tests to address my new additions
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

macite commented on db/schema.rb:
```diff
@@ -26,6 +26,12 @@
   add_index "activity_types", ["abbreviation"], name: "index_activity_types_on_abbreviation", unique: true, using: :btree
   add_index "activity_types", ["name"], name: "index_activity_types_on_name", unique: true, using: :btree
 
+  create_table "ar_internal_metadata", primary_key: "key", force: :cascade do |t|
```
I think this is left over from the rails upgrade...

Do you really need to add this table to the database for this test change to work?

macite replied:
I suggest you revert the changes to this file.

In the future take care with what you are adding to each commit.

macite replied:
These have not been addressed.

macite commented on test/models/learning_outcome_test.rb:
```diff
@@ -0,0 +1,133 @@
+    #Create new unit.
+    learning_outcome_count = LearningOutcome.count
+    unit_1 = {
+        unit_id: 'ASD10001',
```
Unit id is a relation to a unit - not the unit code. This wont work as you expect.

Try adding a simple test that you can actually create a learning outcome and that it is then available within the unit.

macite commented on test/models/learning_outcome_test.rb:
```diff
@@ -0,0 +1,133 @@
+    refute_empty learning_outcome2.abbreviation
+
+    #Check the abbreviation of the unit is unique.
+    refute_equal learning_outcome, learning_outcome2
```
This isn't checking abbreviations...

macite replied:
I'll stop here...

I suggest you make a very simple test first to demonstrate that the basics of learning outcomes are tested. Create the learning outcome and then test that everything about it is setup correctly, and that you can interact with it via its relationships etc.

macite commented on test/models/learning_outcome_test.rb:
```diff
@@ -0,0 +1,133 @@
+
+    learning_outcome2 = LearningOutcome.create!(unit_2)
+    
+    #Check the abbreviation is not empty.
```
Why would it be empty?
Is it valid?
Check details about validations in rails.

macite commented:
Also - please work on your PR description. What you have provided does not tell me what this PR does.

macite commented on test/api/settings_test.rb:
```diff
@@ -0,0 +1,68 @@
+
+    # Get config details
+    def test_get_config_details        
+        expected_ProductName =  Doubtfire::Application.config.institution[:product_name]
```
Variable name case! This is very sloppy.

Guoheng-Liao replied:
Hi Andrew, I have modified variables name case, and unified their format.

macite commented on test/api/settings_test.rb:
```diff
@@ -0,0 +1,68 @@
+    # Get all unit teaching periods info
+    def test_get_all_unit_teaching_periods_info
+        # The GET we are testing
+        get '/api/teaching_periods/id/units'
```
Do we need this API endpoint? It does not sound useful to me. Can you check where it is used in the front end and recommend we delete it if it is not used.

Guoheng-Liao replied:
I have checked the api endpoint, and as you said we don't need it. 
So do I need to delete this link (get '/ api / teaching_periods / id / units') and its test method (def test_get_all_unit_teaching_periods_info) and commit setting_test.rb again?

macite replied:
Sorry, I thought I had replied to this earlier. Yes, please remove the test and the end point.

macite replied:
Once that is done I can merge the PR.

linhhtr commented on test/api/activity_types_api_test.rb:
```diff
@@ -50,4 +50,50 @@ def test_put_activity_types
     assert_json_matches_model(last_response_body, first_activity_type, response_keys)
   end
+
+  def test_post_activity_types_cannot_auth
```
Please change the test name into `test_student_cannot_post_activity_type` for clearer description

linhhtr commented on test/api/activity_types_api_test.rb:
```diff
@@ -50,4 +50,50 @@ def test_put_activity_types
+    assert_equal ActivityType.count, number_of_activity_type
+  end  
+
+  def test_put_activity_types_cannot_auth
```
Also change the name of this test

linhhtr commented on test/api/activity_types_api_test.rb:
```diff
@@ -50,4 +50,50 @@ def test_put_activity_types
+    }
+
+    # Update activity_type with id = 1, but the student user does not have permissions to put it.
+    put_json '/api/activity_types/1', with_auth_token(data_to_put, user)
```
Please do not put `data_to_put` into the activity type of which id is explicitly `1`. Let create a new activity type by FactoryBot `activity_type = FactoryBot.create(:activity_type)` then assign `data_to_put`into to replace that newly created `activity_type` of which id can be called by `activity_type.id`. If this does not make sense to you, please ask

Guoheng-Liao commented:
@macite Hello Andrew, I have updated the variable names in setting_test_api.rb and unified their format.
Also, I added two new unit tests, They are used to test that students cannot post or put new activity types.

Guoheng-Liao commented:
@macite Hello Andrew, I removed the test and code we discussed a while ago.
